### PR TITLE
ADDED: Extended the section on working with tree-shaped data.

### PR DESCRIPTION
**Pull request by wouterbeek**
- Added example data.
- Clarified that the source format is abstracted away from (XML/JSON/other).
- Explained how list elements can be accessed by index.
- Explained how list elements can be iterated over.
- Explained the special meaning of `$index`, `$parent`, and `$root`.
- Added an example where `$parent` != `$root`.
- Explained special value `"__circular"`.

### CODE REVIEW

File: docs/ratt-transform/index.md [hunk 1]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+      {
+        "id": "de",
+        "name": "Germany"
+      }
+    ]
+  }
+}
+```
+
+Since the RATT record abstracts away the syntactic specifics of the source system format, the above example may be read from a JSON source (see [connecting JSON sources](#json)), or from another kind of source.  For example, the RATT record may be read from the following XML source, assuming the contents of the `<root>` tag are selected:
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: This link doesn't work for me.

File: docs/ratt-transform/index.md [hunk 2]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+    </countries>
+    <countries>
+      <id>de</id>
+      <name>Germany</name>
+    </countries>
+  </data>
+</root>
+```
+
+See [the section on connecting XML sources](#xml) for more information.
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: This link goes to github (in the same [page](https://github.com/TriplyDB/Documentation/blob/tree-shaped-data/docs/ratt-transform/index.md) but not to any specific section.

File: docs/ratt-transform/index.md [hunk 3]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+dataset:my-dataset dct:title 'Data about countries.'@en.
+```
+
+<h3 id='accessing-lists-by-index'>Accessing lists by index</h3>
+
+Tree-shaped data formats often allow multiple values to be specified in a sequence.  Examples of this are XML tags with the same key that are directly nested under the same parent, and lists in JSON.
+
+RATT is able to access specific elements from such lists based on their *index* or position.  Following the standard practice in Computer Science, RATT refer to the first element in the list as having index 0.  The second element has index 1, etc.
+
+For the above example record, we can assert the first label of the country:
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: These examples are not about list elements, but objects in an array. It is important to mention that because later RATT works differently with objects that have arrays as values ( `mw.toLiteral.forEach()` and `mw.toIri.forEach()`) and objects with arrays of objects (`mw.forEach()`).

File: docs/ratt-transform/index.md [hunk 4]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+)
+```
+
+This results in the following assertion:
+
+```trig
+country:de rdfs:label 'Germany'@en.
+```
+
+### Iterating over lists
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: Iterating over arrays of objects.

File: docs/ratt-transform/index.md [hunk 5]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+```trig
+country:nl rdfs:label 'The Netherlands'@en-us,
+                      'Nederland'@nl-nl.
+```
+
+The elements that `mw.forEach` iterates over are themselves RATT records.  This implies that all functions that work for full RATT records also work for the RATT records inside `mw.forEach`.
+
+### Special keys when iterating over lists
+
+In the previous section we saw that `mw.forEach` allows us to iterate over the elements in an list.  The elements are themselves made available as RATT records.  The RATT records inside an `mw.forEach `function are smaller.  This allow the regular keys of the iterated-over elements to be accessed directly.
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: 'over objects in an array' instead of 'elements in a list'.

File: docs/ratt-transform/index.md [hunk 6]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+
+The `$root` key is explained in [the next section](#root-key).
+
+<h4 id='root-key'>Root key (`$root`)</h4>
+
+Sometimes it may be necessary to access a part of the original RATT record that is outside of the scope of the `mw.forEach` call.
+
+Every RATT record inside an` mw.forEach` call contains the `"$root"` key.  The value of the root key provides a link to the full RATT record.  Because the `$root` key is part of the linked-to RATT record, it is not possible to print the value of the root key.  (This would result in infinite output.)  For this reason, the value of the `$root` key is printed as the special value `"__circular__"`.
+
+For the above examples, the parent record and root record are the same, but this is not always the case.  Specifically, the parent record and root record are different when `mw.forEach` calls are nested.
```
Comment (dionenak) [inline docs/ratt-transform/index.md:963]: The example of nested forEach is complicated, I find it hard to follow. I would propose to use different data. If we use the above data : 
```
{
  "metadata": {
    "title": {
      "name": "Data about countries."
    }
  },
  "data": {
    "countries": [
      {
        "id": "nl",
        "name": "The Netherlands"
      },
      {
        "id": "de",
        "name": "Germany"
      }
    ]
  }
}
```
and iterate over 'data.countries', the `$parent `would be data but the `$root` will show the whole record including the sibling 'metadata'. Right?

File: docs/ratt-transform/index.md [hunk 7]
```diff
@@ -675,84 +675,384 @@ app.use(
...
+
+### Specifying paths (nested keys)
+
+In tabular data, keys (or column names) are singular.  But in tree-shaped data a *path* of the tree can consist of one or more keys that must be traversed in sequence.
+
+Paths are specified as dot-separated sequences of keys, starting at the top-level and ending at the required value.  For the above example, RATT can access the `"name"` key inside the `"title"` key, which itself is nested inside the `"metadata"` key.  This path is expressed in [1].  Notice that the path expressed in [1] is different from the path expressed in [2], which also access the `"name"` key, but nested inside the `"countries"` and then `"data"` keys.
+
+```
+[1] metadata.title.name
+[2] data.countries.name
```
Comment (dionenak) [inline docs/ratt-transform/index.md]: the second path is not correct: `countries` has an array of objects, while `title` includes an object (which makes path [1]  correct). Thus,
[2] data.countries[x].name where x can be 0 or 1.

File: docs/ratt-transform/index.md (hunk 1 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: I've added section for connecting JSON sources.

File: docs/ratt-transform/index.md (hunk 2 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: This section exists, but on a different page... fixed.

File: docs/ratt-transform/index.md (hunk 3 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md:765]: It is indeed good to think about the terminology here: JSON uses "array" for ordered lists. XML is not very explicit about how ordered lists are called, talking about the plural "elements" inside a parent "element".  I'm dropping my use of "sequence" here, because that term is neither used in JSON nor XML.  Let's go with "ordered list" that is more descriptive and less implementation-specific than "array" or "elements".

I agree that the user should be warned that `forEach` behaves differently on lists of objects than on lists of strings or numbers, but that warning should be part of the documentation for `forEach`.

Comment (wouterbeek) [reply docs/ratt-transform/index.md:765]: Ah, `forEach` follows quickly in the next section I see...

Comment (wouterbeek) [reply docs/ratt-transform/index.md:765]: `mw.toIri.forEach` and `mw.toLiteral.forEach` are not documented yet...

File: docs/ratt-transform/index.md (hunk 4 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md:799]: I'm going to think about this a bit more.  I don't like having different documentation sections for this.

File: docs/ratt-transform/index.md (hunk 6 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md:963]: I agree that nested `mw.forEach` + nested example data is more complicated.

But if we iterate over `'data.countries'` we get records like this (where `$parent`==`$root`):
```
{
  "id": "nl",
  "name": "The Netherlands",
  "$index": 0,
  "$parent": {
    "metadata": {
      "title": {
        "name": "Data about countries."
      }
    },
    "data": {
      "countries": [
        {
          "id": "nl",
          "name": "The Netherlands"
        },
        {
          "id": "de",
          "name": "Germany"
        }
      ]
    }
  },
  "$root": "__circular__"
}
```

I think the rule here is that `$parent` directly contains the first child in the specified sequence.  E.g. `$parent` directly contains `data` in this case.

File: docs/ratt-transform/index.md (hunk 7 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: Ah, you're right.  That breaks my example :-(

I've changed this and refer to the next section where we explain what `[x]` does.

File: docs/ratt-transform/index.md (hunk 3 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: I have added a new section for iterating over lists of primitives.  The section for iterating over lists of objects refers to this new section.

I have made the terminology more consistent: "list" for this kind of data, "array" as the implementation in JSON, "elements" as the implementation in XML.

File: docs/ratt-transform/index.md (hunk 4 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: Renamed the section accordingly.

File: docs/ratt-transform/index.md (hunk 5 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md]: Merged this section with the section called "Iterating over lists of objects".

File: docs/ratt-transform/index.md (hunk 6 above)
Comment (wouterbeek) [reply docs/ratt-transform/index.md:963]: Unfortunately, the more complex example is necessary in this case.  This is left as-is.

**Comment by wouterbeek** [comment]
@dionenak Thanks for the good feedback!  Can you check my latest changes?

File: docs/ratt-transform/index.md (hunk 6 above)
Comment (dionenak) [reply docs/ratt-transform/index.md:963]: you're right! My mistake!
